162 task add try catch blocks in SectionService for robust error handling

CristiSlv1 opened the pull request.

CristiSlv1 committed:
Add SectionServiceException for improved error handling

Introduced a new exception class `SectionServiceException` to manage exceptions in section service operations. Updated multiple methods in the `SectionService` class to include try-catch blocks, enhancing error handling by wrapping exceptions in `SectionServiceException` with descriptive messages. Key methods affected include `GetAllSections`, `GetSectionById`, `GetByRoadmapId`, `CountSectionsFromRoadmap`, `LastOrderNumberFromRoadmap`, `AddSection`, `DeleteSection`, `UpdateSection`, `TrackCompletion`, and `ValidateDependencies`. These changes improve the robustness and clarity of error reporting in the service layer.

CristiSlv1 committed:
Add SectionServiceProxyException for improved error handling

Introduce a new exception class, `SectionServiceProxyException`, in the `Duo.Exceptions` namespace to manage errors related to `SectionServiceProxy` operations. Update multiple methods in the `SectionServiceProxy` class to include try-catch blocks, throwing the new exception with relevant messages when API calls fail. This enhances error handling and provides better context for failures.

UjfalusiAbel commented on Duo/Services/SectionServiceProxy.cs:
```diff
@@ -15,55 +17,125 @@ public SectionServiceProxy(HttpClient httpClient)
-        return await response.Content.ReadFromJsonAsync<int>();
+        try
+        {
+            var response = await httpClient.PostAsJsonAsync("api/sections/add", section);
```
Wrong url, please don't commit while untried

UjfalusiAbel commented on Duo/Services/SectionServiceProxy.cs:
```diff
@@ -15,55 +17,125 @@ public SectionServiceProxy(HttpClient httpClient)
-        return await httpClient.GetFromJsonAsync<int>($"api/sections/count/{roadmapId}");
+        try
+        {
+            return await httpClient.GetFromJsonAsync<int>($"api/sections/count/{roadmapId}");
```
Wrong url, please don't commit while untried

UjfalusiAbel commented on Duo/Services/SectionServiceProxy.cs:
```diff
@@ -15,55 +17,125 @@ public SectionServiceProxy(HttpClient httpClient)
-        await httpClient.DeleteAsync($"api/sections/{sectionId}");
+        try
+        {
+            await httpClient.DeleteAsync($"api/sections/{sectionId}");
```
Wrong url, please don't commit while untried

UjfalusiAbel reviewed (changes requested):
Please fix url problems and next time try if it connects to API not just make a PR

CristiSlv1 committed:
Remove SectionServiceException and SectionServiceProxyException

Deleted the `SectionServiceException` and `SectionServiceProxyException` classes along with their associated `using System;` directives. These classes, which inherited from `Exception`, included constructors for message and inner exception handling.

CristiSlv1 committed:
Update SectionController route to include API prefix

Changed the route attribute for the `SectionController` class from `[Route("section")]` to `[Route("api/section")]`. This update organizes the API structure and clarifies that this controller is part of the API layer.

CristiSlv1 committed:
Refactor SectionServiceProxy for improved error handling

- Moved `SectionServiceProxy` to `Duo.Services` namespace.
- Introduced a `url` variable for API base URL, replacing hardcoded strings.
- Enhanced exception handling with specific logging for `HttpRequestException`.
- Updated methods to use the new `url` variable for API endpoints.
- Removed `SectionServiceProxyException` and logged errors directly to the console.
- Improved overall code structure and readability for better maintenance.

CristiSlv1 committed:
Refactor SectionService for improved error handling

The `SectionService` class has been updated to enhance error handling and refactor section management methods. Key changes include:
- Added `using System.Net.Http;` for HTTP request handling.
- Removed `Duo.Exceptions` to simplify error management.
- Refactored `GetAllSections` to return a list of sections with error handling.
- Updated `AddSection` to validate input and return `0` on failure.
- Improved error handling in `CountSectionsFromRoadmap`, `DeleteSection`, `GetSectionById`, and `LastOrderNumberFromRoadmap`.
- Added logging for errors instead of throwing exceptions in several methods.
- Introduced `TrackCompletion` method for tracking section completion status.
- Enhanced `ValidateDependencies` with error handling.

UjfalusiAbel reviewed (approved):
Approved
